File: Sources/SetlistUI/StageScreen.swift
#if canImport(SwiftUI)
import SwiftUI
import SetlistCore

/// Everything the stage screen draws, in one value. A screen requirement is a
/// resting state, so the view takes state and draws it — it owns no clock, no
/// audio session and no store.
public struct StageState {
    public var model: StageModel
    public var venue: String
    /// Seconds into the set.
    public var elapsed: TimeInterval
    /// Live laugh level, 0...1. Above `speakOverThreshold` the strip is lit.
    public var laughLevel: Double
    public var isRecording: Bool

    public static let speakOverThreshold = 0.35

    public init(
        model: StageModel,
        venue: String,
        elapsed: TimeInterval,
        laughLevel: Double = 0,
        isRecording: Bool = true
    ) {
        self.model = model
        self.venue = venue
        self.elapsed = elapsed
        self.laughLevel = laughLevel
        self.isRecording = isRecording
    }

    var remaining: TimeInterval {
        Countdown.remaining(plannedLength: model.plannedLength, elapsed: elapsed)
    }

    var liveCard: StageCard? {
        model.cards.first(where: { $0.state == .live })
    }
}

/// The live screen. Portrait, one screenful, no scrolling: a comedian scrolling
/// on stage has already lost the room.
public struct StageScreen: View {
    public var state: StageState
    public var onTapCard: (UUID) -> Void
    public var onSetSegmentMode: (SegmentMode) -> Void

    public init(
        state: StageState,
        onTapCard: @escaping (UUID) -> Void = { _ in },
        onSetSegmentMode: @escaping (SegmentMode) -> Void = { _ in }
    ) {
        self.state = state
        self.onTapCard = onTapCard
        self.onSetSegmentMode = onSetSegmentMode
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header
            currentJoke
            segmentModeRow
            setListGrid
        }
        .background(Palette.ink)
        .foregroundStyle(Palette.bone)
    }

    // MARK: - Header

    private var header: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(alignment: .center) {
                HStack(spacing: 8) {
                    if state.isRecording {
                        Circle()
                            .fill(Palette.red)
                            .frame(width: 10, height: 10)
                            .accessibilityLabel("Recording")
                    }
                    Text(state.venue.uppercased())
                        .font(.system(size: 13))
                        .kerning(0.6)
                        .foregroundStyle(Palette.dim)
                }
                Spacer()
                HStack(alignment: .firstTextBaseline, spacing: 7) {
                    Text(Self.countdown(state.remaining))
                        .font(.system(size: 28, weight: .medium, design: .monospaced))
                        .foregroundStyle(countdownColour)
                    Text("\(Self.clock(state.elapsed)) / \(Self.clock(state.model.plannedLength))")
                        .font(.system(size: 14, design: .monospaced))
                        .foregroundStyle(Palette.mute)
                }
            }
            pacingBar
        }
        .padding(.horizontal, 20)
        .padding(.top, 18)
    }

    private var countdownColour: Color {
        switch Countdown.tone(remaining: state.remaining) {
        case .neutral: return Palette.bone
        case .warning: return Palette.orange
        case .over: return Palette.red
        }
    }

    /// One segment per joke, sized by its estimate: the set's shape, and where
    /// in it the comedian actually is.
    private var pacingBar: some View {
        GeometryReader { proxy in
            HStack(spacing: 3) {
                ForEach(state.model.cards) { card in
                    Rectangle()
                        .fill(paceColour(card.state))
                        .frame(width: paceWidth(card, total: proxy.size.width))
                }
            }
        }
        .frame(height: 7)
    }

    private func paceColour(_ cardState: CardState) -> Color {
        switch cardState {
        case .told: return Palette.green
        case .live: return Palette.orange
        case .queued: return Palette.rule
        }
    }

    private func paceWidth(_ card: StageCard, total: CGFloat) -> CGFloat {
        let weights = state.model.cards.map { $0.joke.estimatedLength ?? 120 }
        let sum = weights.reduce(0, +)
        guard sum > 0 else { return 0 }
        let gaps = CGFloat(max(state.model.cards.count - 1, 0)) * 3
        let available = max(total - gaps, 0)
        return available * CGFloat((card.joke.estimatedLength ?? 120) / sum)
    }

    // MARK: - Current joke

    private var currentJoke: some View {
        VStack(alignment: .leading, spacing: 14) {
            if let card = state.liveCard {
                HStack(alignment: .firstTextBaseline) {
                    Text(card.joke.title)
                        .font(.system(size: 32, weight: .medium))
                        .foregroundStyle(.white)
                    Spacer()
                    Text("on \(Self.clock(card.runTime))")
                        .font(.system(size: 16, design: .monospaced))
                        .foregroundStyle(Palette.dim)
                }
                Self.body(of: card.joke)
                    .font(.system(size: 25))
                    .lineSpacing(9)
                if !card.joke.tags.isEmpty || !card.joke.callbacks.isEmpty {
                    tags(for: card.joke)
                }
            } else {
                Text("Tap a card to start")
                    .font(.system(size: 25))
                    .foregroundStyle(Palette.dim)
            }
            laughStrip
        }
        .padding(.horizontal, 22)
        .padding(.vertical, 18)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Palette.panel)
        .overlay(alignment: .leading) { Rectangle().fill(Palette.orange).frame(width: 6) }
        .padding(.horizontal, 16)
        .padding(.top, 16)
    }

    /// The joke's text with its crucial setups highlighted — the words the
    /// punch depends on, the ones a tired comic paraphrases away.
    static func body(of joke: Joke) -> Text {
        let characters = Array(joke.body)
        let marked = joke.setups
            .filter { $0.start >= 0 && $0.length > 0 && $0.start + $0.length <= characters.count }
            .sorted { $0.start < $1.start }

        var result = Text("")
        var cursor = 0
        for span in marked where span.start >= cursor {
            if span.start > cursor {
                result = result + Text(String(characters[cursor ..< span.start])).foregroundColor(Palette.bone)
            }
            let highlighted = String(characters[span.start ..< (span.start + span.length)])
            result = result + Text(highlighted).foregroundColor(Palette.ink).bold()
            cursor = span.start + span.length
        }
        if cursor < characters.count {
            result = result + Text(String(characters[cursor...])).foregroundColor(Palette.bone)
        }
        return result
    }

    private func tags(for joke: Joke) -> some View {
        HStack(spacing: 8) {
            ForEach(joke.callbacks, id: \.self) { _ in
                tag("callback", accent: Palette.green)
            }
            ForEach(joke.tags, id: \.self) { text in
                tag(text, accent: Palette.dim)
            }
        }
    }

    private func tag(_ text: String, accent: Color) -> some View {
        Text(text)
            .font(.system(size: 15))
            .foregroundStyle(accent)
            .padding(.horizontal, 12)
            .padding(.vertical, 4)
            .overlay(Rectangle().stroke(accent.opacity(0.5), lineWidth: 1))
    }

    /// One bar, lit while the room is louder than the comedian can talk over.
    /// The one thing genuinely hard to judge from behind a mic.
    private var laughStrip: some View {
        Rectangle()
            .fill(state.laughLevel > StageState.speakOverThreshold ? Palette.green : Palette.rule)
            .frame(height: 10)
            .frame(maxWidth: .infinity)
    }

    // MARK: - Segment mode

    private var segmentModeRow: some View {
        HStack(spacing: 13) {
            Toggle("", isOn: Binding(
                get: { state.model.segmentMode == .autodetect },
                set: { onSetSegmentMode($0 ? .autodetect : .manual) }
            ))
            .labelsHidden()
            .disabled(!state.model.autodetectAvailable)
            Text(state.model.segmentMode == .autodetect ? "Autodetect bits" : "Manual switch")
                .font(.system(size: 17))
                .foregroundStyle(state.model.segmentMode == .autodetect ? Palette.green : Palette.dim)
            Spacer()
        }
        .padding(.horizontal, 16)
        .padding(.top, 16)
    }

    // MARK: - Set list

    private var setListGrid: some View {
        VStack(alignment: .leading, spacing: 9) {
            HStack {
                Text("SET LIST · \(state.model.startedCount) / \(state.model.plannedCount)")
                    .font(.system(size: 14))
                    .kerning(0.6)
                Spacer()
                HStack(spacing: 13) {
                    key("told", Palette.green)
                    key("live", Palette.orange)
                    key("queued", Palette.edge)
                }
            }
            .foregroundStyle(Palette.dim)

            LazyVGrid(
                columns: Array(repeating: GridItem(.flexible(), spacing: 9), count: 3),
                spacing: 9
            ) {
                ForEach(state.model.cards) { card in
                    Button { onTapCard(card.id) } label: { cardFace(card) }
                        .buttonStyle(.plain)
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.top, 14)
        .padding(.bottom, 16)
        .frame(maxHeight: .infinity, alignment: .top)
    }

    private func key(_ label: String, _ colour: Color) -> some View {
        HStack(spacing: 5) {
            Rectangle().fill(colour).frame(width: 10, height: 10)
            Text(label).font(.system(size: 13))
        }
    }

    /// A card carries its joke and how long it ran — never a laugh score.
    /// Requirement 2.7: scores off the stage screen.
    private func cardFace(_ card: StageCard) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(card.joke.title)
                .font(.system(size: 22, weight: .medium))
                .foregroundStyle(cardTitleColour(card.state))
                .multilineTextAlignment(.leading)
            Spacer(minLength: 0)
            Text(cardFootnote(card))
                .font(.system(size: 14, design: card.state == .queued ? .default : .monospaced))
                .foregroundStyle(cardFootnoteColour(card.state))
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 13)
        .frame(maxWidth: .infinity, minHeight: 96, alignment: .leading)
        .background(cardBackground(card.state))
    }

    private func cardFootnote(_ card: StageCard) -> String {
        switch card.state {
        case .live: return "LIVE · \(Self.clock(card.runTime))"
        case .told: return Self.clock(card.runTime)
        case .queued: return card.joke.estimatedLength.map { "~\(Self.clock($0))" } ?? "—"
        }
    }

    private func cardBackground(_ cardState: CardState) -> Color {
        switch cardState {
        case .told: return Palette.greenDeep
        case .live: return Palette.orange
        case .queued: return Palette.panel
        }
    }

    private func cardTitleColour(_ cardState: CardState) -> Color {
        switch cardState {
        case .told: return Palette.green
        case .live: return Palette.ink
        case .queued: return Palette.bone
        }
    }

    private func cardFootnoteColour(_ cardState: CardState) -> Color {
        switch cardState {
        case .told: return Palette.greenInk
        case .live: return Palette.orangeDeep
        case .queued: return Palette.dim
        }
    }

    // MARK: - Formatting

    /// `m:ss`, the only time format on the screen.
    static func clock(_ seconds: TimeInterval) -> String {
        let whole = Int(seconds.rounded(.down))
        return "\(whole / 60):\(String(format: "%02d", whole % 60))"
    }

    /// The countdown reads as a negative once the slot is over.
    static func countdown(_ remaining: TimeInterval) -> String {
        remaining < 0 ? "+\(clock(-remaining))" : "−\(clock(remaining))"
    }
}
#endif

File: packages/setlist_ui/lib/src/stage_screen.dart
import 'package:flutter/widgets.dart';
import 'package:flutter/material.dart' show Material, MaterialType, Switch;
import 'package:setlist_core/setlist_core.dart';

import 'formatting.dart';
import 'palette.dart';
import 'stage_region.dart';

/// Everything the stage screen draws, in one value.
///
/// A screen requirement is a resting state, so the view takes state and draws
/// it — it owns no clock, no audio session and no store.
class StageState {
  const StageState({
    required this.model,
    required this.venue,
    required this.elapsed,
    this.laughLevel = 0,
    this.isRecording = true,
  });

  final StageModel model;
  final String venue;

  /// How far into the set.
  final Duration elapsed;

  /// Live laugh level, 0..1. Above [speakOverThreshold] the strip is lit.
  final double laughLevel;
  final bool isRecording;

  static const double speakOverThreshold = 0.35;

  Duration get remaining =>
      Countdown.remaining(plannedLength: model.plannedLength, elapsed: elapsed);

  StageCard? get liveCard {
    for (final card in model.cards) {
      if (card.state == CardState.live) return card;
    }
    return null;
  }

  /// A card's run time as of now — the live joke's keeps counting.
  Duration runTimeOf(StageCard card) =>
      model.runTimeOf(card.id, at: CaptureTime(elapsed.inMicroseconds / 1e6));
}

/// The live screen.
///
/// One screenful, no scrolling region: a comedian scrolling on stage has
/// already lost the room.
class StageScreen extends StatelessWidget {
  const StageScreen({
    super.key,
    required this.state,
    this.onTapCard,
    this.onSetSegmentMode,
  });

  final StageState state;
  final void Function(String jokeId)? onTapCard;
  final void Function(SegmentMode mode)? onSetSegmentMode;

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Palette.ink,
      // A transparent Material, purely so the switch has the ancestor its
      // implementation demands. It paints nothing of its own: the ink above is
      // still what the screen is drawn on.
      child: Material(
        type: MaterialType.transparency,
        child: DefaultTextStyle(
        style: const TextStyle(
          fontFamily: Typeface.sans,
          color: Palette.bone,
          fontSize: 15,
        ),
          child: LayoutBuilder(
            builder: (context, constraints) {
              final narrow = constraints.maxWidth < narrowWidth;
              // The panel is capped when the screen is short *or* narrow, and
              // the two catch different devices: a portrait phone is narrow but
              // tall, a landscape phone is wide but short, and on both the
              // panel would otherwise take the height the set list needs.
              final short = constraints.maxHeight < shortHeight;
              final capped = narrow || short;
              return Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  _header(),
                  _currentJoke(capped: capped, available: constraints.maxHeight),
                  _segmentModeRow(short: short),
                  Expanded(child: _setListGrid(narrow: narrow, short: short)),
                ],
              );
            },
          ),
        ),
      ),
    );
  }

  // ---------------------------------------------------------------- header

  Widget _header() {
    return Padding(
      padding: const EdgeInsets.fromLTRB(20, 18, 20, 0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              // The room's name gives way first when the screen is narrow: on a
              // phone the countdown and the elapsed pair have to stay whole,
              // and a comedian already knows which room they are in.
              Flexible(
                child: Row(
                  key: StageRegion.venue.key,
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    if (state.isRecording) ...[
                      Container(
                        width: 10,
                        height: 10,
                        decoration: const BoxDecoration(
                          color: Palette.red,
                          shape: BoxShape.circle,
                        ),
                      ),
                      const SizedBox(width: 8),
                    ],
                    Flexible(
                      child: Text(
                        state.venue.toUpperCase(),
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: const TextStyle(
                          fontSize: 13,
                          letterSpacing: 0.6,
                          color: Palette.dim,
                        ),
                      ),
                    ),
                  ],
                ),
              ),
              const SizedBox(width: 12),
              // Not flexible: the clocks are the part that must stay whole, so
              // the room's name is the only thing that yields.
              Row(
                key: StageRegion.clocks.key,
                mainAxisSize: MainAxisSize.min,
                crossAxisAlignment: CrossAxisAlignment.baseline,
                textBaseline: TextBaseline.alphabetic,
                children: [
                  Text(
                    countdown(state.remaining),
                    style: TextStyle(
                      fontFamily: Typeface.mono,
                      fontSize: 28,
                      color: _countdownColour,
                    ),
                  ),
                  const SizedBox(width: 7),
                  Text(
                    '${clock(state.elapsed)} / ${clock(state.model.plannedLength)}',
                    style: const TextStyle(
                      fontFamily: Typeface.mono,
                      fontSize: 14,
                      color: Palette.mute,
                    ),
                  ),
                ],
              ),
            ],
          ),
          const SizedBox(height: 12),
          _pacingBar(),
        ],
      ),
    );
  }

  Color get _countdownColour => switch (Countdown.tone(state.remaining)) {
        CountdownTone.neutral => Palette.bone,
        CountdownTone.warning => Palette.orange,
        CountdownTone.over => Palette.red,
      };

  /// One segment per joke, sized by its estimate: the set's shape, and where in
  /// it the comedian actually is.
  Widget _pacingBar() {
    return SizedBox(
      key: StageRegion.pacingBar.key,
      height: 7,
      child: LayoutBuilder(
        builder: (context, constraints) {
          final cards = state.model.cards;
          // A joke with no estimate still has to occupy the bar, or the set's
          // shape would silently omit it. Two minutes is the drawing's stand-in
          // and nothing else reads it — the planned-length rule in the core
          // keeps treating an unestimated joke as contributing nothing.
          double weightOf(StageCard card) =>
              (card.joke.estimatedLength ?? const Duration(seconds: 120)).inMicroseconds / 1e6;
          final sum = cards.fold<double>(0, (total, card) => total + weightOf(card));
          if (sum <= 0) return const SizedBox.shrink();
          final gaps = (cards.length - 1).clamp(0, cards.length) * 3.0;
          final available = (constraints.maxWidth - gaps).clamp(0.0, double.infinity);
          return Row(
            // Stretch, so each segment takes the bar's full height: a box with
            // only a width and a colour has no height of its own, and the bar
            // renders as nothing at all.
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              for (var i = 0; i < cards.length; i++) ...[
                if (i > 0) const SizedBox(width: 3),
                SizedBox(
                  width: available * weightOf(cards[i]) / sum,
                  child: ColoredBox(color: _paceColour(cards[i].state)),
                ),
              ],
            ],
          );
        },
      ),
    );
  }

  static Color _paceColour(CardState state) => switch (state) {
        CardState.told => Palette.green,
        CardState.live => Palette.orange,
        CardState.queued => Palette.rule,
      };

  // ---------------------------------------------------------- current joke

  static Widget _flexIf(bool flex, Widget child) =>
      flex ? Flexible(child: child) : child;

  /// The share of the screen the joke panel may take before the set list
  /// starts losing cards it cannot afford to lose.
  ///
  /// Only applied on a narrow screen. A phone cannot show the whole body *and*
  /// nine readable cards, so the body is what gives: a comedian mid-bit is
  /// reading the top of it, and the grid is what they navigate by.
  static const double cappedPanelShare = 0.34;

  /// The height below which the joke panel has to give the set list room.
  static const double shortHeight = 600;

  Widget _currentJoke({required bool capped, required double available}) {
    final card = state.liveCard;
    return Padding(
      padding: const EdgeInsets.fromLTRB(16, 16, 16, 0),
      child: ConstrainedBox(
        constraints: BoxConstraints(
          maxHeight: capped ? available * cappedPanelShare : double.infinity,
        ),
        child: Container(
        key: StageRegion.currentJoke.key,
        width: double.infinity,
        color: Palette.panel,
        child: Stack(
          children: [
            Positioned(left: 0, top: 0, bottom: 0, width: 6,
                child: const ColoredBox(color: Palette.orange)),
            Padding(
              // A capped panel tightens its own rhythm before it gives up any
              // of its content: the title, the tags and the strip all still
              // have to be there, and only the space between them is spare.
              padding: EdgeInsets.symmetric(
                horizontal: 22,
                vertical: capped ? 10 : 18,
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  if (card != null) ...[
                    Row(
                      key: StageRegion.jokeTitleRow.key,
                      crossAxisAlignment: CrossAxisAlignment.baseline,
                      textBaseline: TextBaseline.alphabetic,
                      children: [
                        Expanded(
                          child: Text(
                            card.joke.title,
                            style: const TextStyle(
                              fontSize: 32,
                              fontWeight: FontWeight.w700,
                              color: Color(0xFFFFFFFF),
                            ),
                          ),
                        ),
                        Text(
                          'on ${clock(state.runTimeOf(card))}',
                          style: const TextStyle(
                            fontFamily: Typeface.mono,
                            fontSize: 16,
                            color: Palette.dim,
                          ),
                        ),
                      ],
                    ),
                    const SizedBox(height: 14),
                    // Flexible only where there is a bound to flex within: on
                    // a wide screen the panel is unbounded, and a flex child in
                    // an unbounded column is a layout error rather than a
                    // shrink-wrap.
                    _flexIf(
                      capped,
                      Text.rich(
                        jokeBody(card.joke),
                        key: StageRegion.jokeBody.key,
                        overflow: capped ? TextOverflow.ellipsis : TextOverflow.clip,
                        style: const TextStyle(fontSize: 25, height: 1.36),
                      ),
                    ),
                    if (card.joke.tags.isNotEmpty || card.joke.callbacks.isNotEmpty) ...[
                      SizedBox(height: capped ? 8 : 14),
                      _tags(card.joke),
                    ],
                  ] else
                    Text(
                      'Tap a card to start',
                      key: StageRegion.jokeBody.key,
                      style: const TextStyle(fontSize: 25, color: Palette.dim),
                    ),
                  SizedBox(height: capped ? 8 : 14),
                  _laughStrip(),
                ],
              ),
            ),
          ],
        ),
        ),
      ),
    );
  }

  /// The joke's text with its crucial setups highlighted — the words the punch
  /// depends on, the ones a tired comic paraphrases away.
  ///
  /// The highlight is weight and the live colour, never a darker shade, which
  /// reads as de-emphasis. That is requirement 3.3's wording rather than a
  /// platform limit: unlike the SwiftUI original, a span here could carry a
  /// background, and the spec is what says it should not.
  static TextSpan jokeBody(Joke joke) {
    final characters = joke.body.split('');
    final marked = [
      for (final span in joke.setups)
        if (span.start >= 0 && span.length > 0 && span.start + span.length <= characters.length)
          span,
    ]..sort((a, b) => a.start.compareTo(b.start));

    final spans = <TextSpan>[];
    var cursor = 0;
    for (final span in marked) {
      if (span.start < cursor) continue; // overlapping spans: first one wins
      if (span.start > cursor) {
        spans.add(TextSpan(
          text: characters.sublist(cursor, span.start).join(),
          style: const TextStyle(color: Palette.bone),
        ));
      }
      spans.add(TextSpan(
        text: characters.sublist(span.start, span.start + span.length).join(),
        style: const TextStyle(color: Palette.orange, fontWeight: FontWeight.w700),
      ));
      cursor = span.start + span.length;
    }
    if (cursor < characters.length) {
      spans.add(TextSpan(
        text: characters.sublist(cursor).join(),
        style: const TextStyle(color: Palette.bone),
      ));
    }
    return TextSpan(children: spans);
  }

  Widget _tags(Joke joke) {
    // Wraps rather than overflowing: the tags are the panel's least urgent
    // content, so on a narrow screen they take a second line instead of
    // pushing themselves off the edge.
    return Wrap(
      key: StageRegion.jokeTags.key,
      spacing: 8,
      runSpacing: 8,
      children: [
        for (final _ in joke.callbacks) _tag('callback', Palette.green),
        for (final text in joke.tags) _tag(text, Palette.dim),
      ],
    );
  }

  Widget _tag(String text, Color accent) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 4),
      decoration: BoxDecoration(
        border: Border.all(color: accent.withValues(alpha: 0.5)),
      ),
      child: Text(text, style: TextStyle(fontSize: 15, color: accent)),
    );
  }

  /// One bar, lit while the room is louder than the comedian can talk over.
  /// The one thing genuinely hard to judge from behind a mic.
  Widget _laughStrip() {
    return Container(
      key: StageRegion.laughStrip.key,
      height: 10,
      width: double.infinity,
      color: state.laughLevel > StageState.speakOverThreshold ? Palette.green : Palette.rule,
    );
  }

  // ---------------------------------------------------------- segment mode

  Widget _segmentModeRow({required bool short}) {
    final on = state.model.segmentMode == SegmentMode.autodetect;
    return Padding(
      // A short screen buys the set list its last rows out of the spacing
      // between blocks before it takes anything out of the blocks themselves.
      padding: EdgeInsets.fromLTRB(16, short ? 6 : 16, 16, 0),
      child: Row(
        key: StageRegion.segmentMode.key,
        children: [
          Switch(
            value: on,
            onChanged: state.model.autodetectAvailable
                ? (value) => onSetSegmentMode?.call(
                    value ? SegmentMode.autodetect : SegmentMode.manual)
                : null,
          ),
          const SizedBox(width: 13),
          Text(
            on ? 'Autodetect bits' : 'Manual switch',
            style: TextStyle(fontSize: 17, color: on ? Palette.green : Palette.dim),
          ),
        ],
      ),
    );
  }

  // -------------------------------------------------------------- set list

  Widget _setListGrid({required bool narrow, required bool short}) {
    return Padding(
      padding: EdgeInsets.fromLTRB(16, short ? 6 : 14, 16, short ? 6 : 16),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          // The count and the key share a line where there is room and stack
          // where there is not. Both have to be on the screen — 4.4 and 4.3 —
          // and a narrow phone cannot hold them side by side.
          Wrap(
            alignment: WrapAlignment.spaceBetween,
            spacing: 13,
            runSpacing: 6,
            children: [
              SizedBox(
                width: narrow ? double.infinity : null,
                child: Text(
                  'SET LIST · ${state.model.startedCount} / ${state.model.plannedCount}',
                  key: StageRegion.setListCount.key,
                  style: const TextStyle(fontSize: 14, letterSpacing: 0.6, color: Palette.dim),
                ),
              ),
              Row(
                key: StageRegion.setListKey.key,
                mainAxisSize: MainAxisSize.min,
                children: [
                  _key('told', Palette.green),
                  const SizedBox(width: 13),
                  _key('live', Palette.orange),
                  const SizedBox(width: 13),
                  _key('queued', Palette.edge),
                ],
              ),
            ],
          ),
          const SizedBox(height: 9),
          // Rows rather than a scrolling grid: the set list takes whatever
          // height the joke panel leaves, so the cards fill one screen instead
          // of sizing to their text and stranding the bottom third.
          Expanded(
            child: Column(
              key: StageRegion.grid.key,
              children: [
                for (var r = 0; r < _rowsOf(narrow ? 2 : 3).length; r++) ...[
                  if (r > 0) const SizedBox(height: 9),
                  Expanded(
                    child: Row(
                      children: [
                        for (var c = 0; c < (narrow ? 2 : 3); c++) ...[
                          if (c > 0) const SizedBox(width: 9),
                          Expanded(
                            child: c < _rowsOf(narrow ? 2 : 3)[r].length
                                ? _cardFace(_rowsOf(narrow ? 2 : 3)[r][c])
                                : const SizedBox.shrink(),
                          ),
                        ],
                      ],
                    ),
                  ),
                ],
              ],
            ),
          ),
        ],
      ),
    );
  }

  /// The width below which the set list drops to two columns.
  ///
  /// Three cards across a phone leaves each about 120pt, which clips a title
  /// like "Flatmate rota" to five characters, and a card nobody can read is
  /// worse than a taller grid.
  static const double narrowWidth = 600;

  List<List<StageCard>> _rowsOf(int columns) {
    final cards = state.model.cards;
    return [
      for (var start = 0; start < cards.length; start += columns)
        cards.sublist(start, start + columns > cards.length ? cards.length : start + columns),
    ];
  }

  Widget _key(String label, Color colour) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        Container(width: 10, height: 10, color: colour),
        const SizedBox(width: 5),
        Text(label, style: const TextStyle(fontSize: 13, color: Palette.dim)),
      ],
    );
  }

  /// A card carries its joke and how long it ran — never a laugh score
  /// (requirement 4.6).
  Widget _cardFace(StageCard card) {
    return GestureDetector(
      onTap: () => onTapCard?.call(card.id),
      child: Container(
        color: _cardBackground(card.state),
        padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 13),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Expanded(
              child: Text(
                card.joke.title,
                // Wraps and then ellipsises rather than cutting mid-word: a
                // title that reads "Flatmate" is a different bit from one that
                // reads "Flatmate rota", and the comedian is scanning for the
                // one they mean.
                maxLines: 2,
                overflow: TextOverflow.ellipsis,
                style: TextStyle(
                  fontSize: 22,
                  fontWeight: FontWeight.w700,
                  color: _cardTitleColour(card.state),
                ),
              ),
            ),
            Text(
              _cardFootnote(card),
              style: TextStyle(
                fontFamily: card.state == CardState.queued ? Typeface.sans : Typeface.mono,
                fontSize: 14,
                color: _cardFootnoteColour(card.state),
              ),
            ),
          ],
        ),
      ),
    );
  }

  String _cardFootnote(StageCard card) => switch (card.state) {
        CardState.live => 'LIVE · ${clock(state.runTimeOf(card))}',
        CardState.told => clock(state.runTimeOf(card)),
        CardState.queued => card.joke.estimatedLength == null
            ? '—'
            : '~${clock(card.joke.estimatedLength!)}',
      };

  static Color _cardBackground(CardState state) => switch (state) {
        CardState.told => Palette.greenDeep,
        CardState.live => Palette.orange,
        CardState.queued => Palette.panel,
      };

  static Color _cardTitleColour(CardState state) => switch (state) {
        CardState.told => Palette.green,
        CardState.live => Palette.ink,
        CardState.queued => Palette.bone,
      };

  static Color _cardFootnoteColour(CardState state) => switch (state) {
        CardState.told => Palette.greenInk,
        CardState.live => Palette.orangeDeep,
        CardState.queued => Palette.dim,
      };
}
